elks/arch/i86/kernel/irq.c

```diff
@@ -72,6 +73,42 @@ static void default_handler(int i, void *regs, void *dev)
 	printk("Unexpected interrupt: %u\n", i);
 }
 
+
+typedef void (* int_proc) (void);
+
+struct int_handler {
+	byte_t call;  // CALLF opcode (9Ah)
+	int_proc proc;
+	word_t seg;
+	byte_t irq;
+} __attribute__ ((packed));
+
+typedef struct int_handler int_handler_s;
+
+// TODO: move to irq.h
+int int_vector_set (int vect, int_proc proc, int seg);
+void _irqit (void);
+
+// TODO: move to segment.h
+int seg_data (void);
+
+// Add a dynamically allocated handler
+// that redirects to the generic handler
+
+static int int_handler_add (int irq, int vect)
+	{
+	int_handler_s * h = (int_handler_s *) heap_alloc (sizeof (int_handler_s), HEAP_TAG_INTHAND);
+	if (!h) return -ENOMEM;
+
+	h->call = 0x9A;    // CALLF opcode
```

Comment:
0x9A adds architecture-specific code back into this routine. Consider calling `int_vector_set(irq, h)` and having it handle all architecture-specific code in irqtab.S, which is not cleaned up. If rather stay some in C, then this routine needs to move to irq-8259.c (except then it needs duplicating in irq-8018x.c, so we're not solving our cleanup.)

Also, now every interrupt costs extra far call...

Reply:
You are in the folder `arch/i86`, mister :laughing:

Reply:
Yes, one more segment push on each interrupt... wow... what a big performance impact :smirk: !

Reply:
Hey, it all adds up. I am still considering what is the true benefit of using dynamic approach, as well as configuration issue brought up in https://github.com/jbruchon/elks/pull/940#issuecomment-841700174. What do you think? This approach, which I am for, but only cautiously, is already bigger than anticipated by both of us, and certainly will have to go through hardware and device testing before commit.

Reply:
Approach exposed in the main thread.